pycrostates/preprocessing/spatial_filter.py

```diff
@@ -0,0 +1,165 @@
+    inst: Union[BaseRaw, BaseEpochs],
+    ch_type: str = "eeg",
+    exclude_bads: bool = True,
+    n_jobs: int = 1,
+    verbose=None,
+):
+    r"""Apply spatial filter.
+
+    Adapted from \ :footcite:p:`michel2019eeg`.
+    Apply an instantaneous filter which interpolates channels
+    with local neigbhors while removing outliers.
+    The current implementation proceeds as follows:
+
+    * An interpolation matrix is computed using
+      ``mne.channels.interpolation._make_interpolation_matrix``.
+    * An ajdacency matrix is computed using
+      `mne.channels.find_ch_adjacency`.
+    * If ``exclude_bads`` is st to ``True``,
+      bad channels are removed from the ajdacency matrix.
+    * For each timepoint and each channel,
+      a reduced adjacency matrix is build by removing neigbhors
+      with lowest and highest value.
+    * For each timepoint and each channel,
+      a reduced interpolation matrix is build by extracting neigbhor
+      weights based on the reduced adjacency matrix.
+    * Reduced interpolation matrices are normalized.
+    * Channel's timepoints are interpolated
+      using their reduced interpolation matrix.
+
+    Parameters
+    ----------
+    inst : Raw | Epochs
+        Instance to spatial filter.
+    ch_type : str
+        The channel type on which to apply spatial filter.
+        Currently only supports ``'eeg'``.
+        Default to 'eeg'.
+    exclude_bads : bool
+        If set to ``True`` bad channels will be removed
+        from the adjacency matrix and therefore no be used
+        to interpolate neighbors. In addition, bad channels
+        will not be filtered.
+        If set to ``False``, proceed as if all channels were good.
+        Default to ``True``.
+    %(n_jobs)s
+    %(verbose)s
+
+    Returns
+    -------
+    inst : Raw | Epochs
+        The modified instance.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    # Checks
+    _check_type(inst, (BaseRaw, BaseEpochs))
+    _check_value(ch_type, ("eeg"), item_name="ch_type")
+    n_jobs = _check_n_jobs(n_jobs)
+    # check preload for Raw
+    _check_preload(inst, "Apply spatial filter")
+    # remove bad channels
+    # Extract adjacency matrix
+    adjacency_matrix, ch_names = mne.channels.find_ch_adjacency(
+        inst.info, ch_type
+    )
+    adjacency_matrix = adjacency_matrix.todense()
+    adjacency_matrix = np.array(adjacency_matrix)
+    if exclude_bads:
+        for c, chan in enumerate(ch_names):
+            if chan in inst.info["bads"]:
+                adjacency_matrix[c, :] = 0  # do not change bads
+                adjacency_matrix[:, c] = 0  # don't use bads to interpolate
+                print(adjacency_matrix)
+    # retrieve picks based on adjacency matrix
+    picks = dict(_picks_by_type(inst.info, exclude=[]))[ch_type]
+    assert ch_names == inst.ch_names
+    # retrieve channel positions
+    # TODO: warn if no sphere fit for eeg
+    pos = inst._get_channel_positions(picks)
```

Comment:
If you need a montage, we should check at the beginning that the instance does have one: `inst.get_montage() is not None`.